### PR TITLE
Add 'new' omega score term. 

### PR DESCRIPTION
**Pull request by jflat06**
Initial commit just to get things in. Will add full functionality as implemented.

### CODE REVIEW

**Comment by jflat06** [comment]
Getting some non-deterministic behavior.

Sometimes the first value in the 10-pose energies will not match any of the rest of them, and will cause the test to fail. Other times, the GPU portion segfaults. I'm getting these are related.

**Comment by jflat06** [comment]
Suspecting that the non-deterministic behavior is happening at the last residue, since I do no checks to see if there is anything on the other end of the connection.

File: tmol/score/omega/params.py
Comment (aleaverfay) [inline tmol/score/omega/params.py:26]: FWIW, the "ParamResolver" idiom is mostly a holdover from the PackedResidueSystem way of approaching scoring. The ParamResolvers would take a set of residues and create denormalized tensors of parameters so that each atom had its own parameters and didn't need to be thought of as part of a particular residue. I don't believe you need a param resolver for omega; the tensor of the GlobalParams can be created simply by the OmegaEnergyTerm in its __init__ method.

File: tmol/score/omega/potentials/omega.hh
Comment (aleaverfay) [inline tmol/score/omega/potentials/omega.hh:16]: I would like to avoid "using namespace" statements in header files

File: tmol/score/omega/potentials/compiled.ops.cpp
Comment (jflat06) [reply tmol/score/omega/potentials/compiled.ops.cpp:143]: Hey Andrew, this code is very much unfinished - I was mostly checking it in to have somewhere to discuss issues/questions. I haven't done anything for the backward step at all yet, this is just copy pasted code from LJLK (I think).

File: tmol/score/omega/potentials/omega_pose_score.hh
Comment (aleaverfay) [inline tmol/score/omega/potentials/omega_pose_score.hh:47]: The convention for pose scores is that each term should return a 2D tensor of values:
1st dimension: the number of sub-terms defined (e.g. lk-ball defines 4, fa_elec defines 1)
2nd dimension: the number of poses in the stack.
That way the set of 2D tensors can be concatenated along their first dimensions and then the weighted scores  computed by a matrix-vector dot product (see tmol/score/score_function.py:186)

Comment (jflat06) [reply tmol/score/omega/potentials/omega_pose_score.hh:47]: Ok, I was wondering about that, as the shape of the baseline I copied from LJLK didn't match. Thanks!

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (aleaverfay) [inline tmol/score/omega/potentials/omega_pose_score.impl.hh:70]: Yes, this may be -1 if there is no residue on the other side of a connection. This is actually how we handle termini right now. This function looks a lot like a function I've sketched (but not tested) for Rama. 

https://github.com/uw-ipd/tmol/blob/aleaverfay/res_centric_rama/tmol/score/rama/potentials/rama_pose_score.impl.hh#L174 

It would probably be wise to refactor both of these functions into a common location. In any case, the function should return -1 for any uaid that failed to resolve and then the calling function would need to look for that condition.

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (jflat06) [inline tmol/score/omega/potentials/omega_pose_score.impl.hh:81]: @aleaverfay This is where I suspect things might be producing non-deterministic behavior since I do nothing to check if a connection is valid.

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (jflat06) [reply tmol/score/omega/potentials/omega_pose_score.impl.hh:70]: Ok, great!

Is there a reason to prefer lambdas over declaring a function in the namespace like this?

Comment (jflat06) [reply tmol/score/omega/potentials/omega_pose_score.impl.hh:70]: Any suggestions on a proper home for this code?

File: tmol/score/omega/potentials/omega.hh
Comment (jflat06) [reply tmol/score/omega/potentials/omega.hh:16]: Huh, I think I coped it from another header file that I was using as an example. I've moved it to compiled.ops.cpp where it is needed.

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (jflat06) [reply tmol/score/omega/potentials/omega_pose_score.impl.hh:70]: Handling the -1 case _appears_ to have resolved the non-deterministic behavior, at least from my ~15 manual runs.

File: tmol/score/omega/omega_energy_term.py
Comment (jflat06) [inline tmol/score/omega/omega_energy_term.py:16]: Is this actual the correct usage of AtomTypeDependentTerm and BondDependentTerm? When are these necessary/what do they provide?

File: tmol/score/omega/omega_energy_term.py
Comment (jflat06) [inline tmol/score/omega/omega_energy_term.py:34]: Technically the calculation involves 2 residues, but it is O(n) on the number of blocks. What should this be?

File: tmol/score/omega/omega_energy_term.py
Comment (jflat06) [inline tmol/score/omega/omega_energy_term.py:42]: Should I be putting these in a device torch tensor here? I think it's probably unnecessary because the setup_packed_block_types func makes a copy into a separate tensor regardless, and only that tensor is passed down into the kernel.

What is the purpose of the distinction between setup_block_type and setup_packed_block_types, then? Why not just do everything in setup_packed_block_types?

File: tmol/score/omega/params.py
Comment (jflat06) [inline tmol/score/omega/params.py:36]: This is basically just packing the OmegaDatabase into tensors so they're available on the GPU. Does this look right?

From here, the OmegaWholePoseScoringModule labels these as parameters with the line: 

`self.global_params = _p(torch.stack(_t([global_params.K]), dim=1))`

which concatenates them all into a single tensor? Is this more complicated than it needs to be?

File: tmol/score/omega/potentials/compiled.ops.cpp
Comment (jflat06) [inline tmol/score/omega/potentials/compiled.ops.cpp:116]: Remove before commit - was useful for getting printf() output from GPU.

File: tmol/score/omega/potentials/params.hh
Comment (jflat06) [inline tmol/score/omega/potentials/params.hh:76]: I just copy/pasted this from LJLK, and don't really know what's going on here. Are there any specific considerations? Is this code that can be factored out somehow?

File: tmol/tests/score/omega/test_omega_energy_term.py
Comment (jflat06) [inline tmol/tests/score/omega/test_omega_energy_term.py:40]: How are these parameters? Any tweaks I should make?

File: tmol/tests/score/omega/test_omega_energy_term.py
Comment (jflat06) [inline tmol/tests/score/omega/test_omega_energy_term.py:10]: What is the standard set of tests that I should implement?

File: tmol/tests/score/omega/test_omega_energy_term.py
Comment (jflat06) [inline tmol/tests/score/omega/test_omega_energy_term.py:16]: Anything else I should test in here?

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (jflat06) [inline tmol/score/omega/potentials/omega_pose_score.impl.hh:121]: What do these do? Are these appropriate for omega?

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (jflat06) [inline tmol/score/omega/potentials/omega_pose_score.impl.hh:168]: Is this the correct dispatch that I should be using for omega?

**Comment by jflat06** [comment]
I have implemented the following tests:

test_smoke
test_whole_pose_scoring_module_gradcheck_whole_pose
test_whole_pose_scoring_module_10

Are there other standard tests that I should include as well? Anything specific to omega?

**Comment by jflat06** [comment]
@aleaverfay @fdimaio I've cleaned up the code and made a few comments/questions, if you have some time to take a look.

**Review by aleaverfay — CHANGES_REQUESTED** [review]
This is looking really good. After you've cleaned up the linting (let me know if any part of that process is unclear), the handful of small changes I suggested, and fixed the "remove before merging" cudaDeviceSynchronize call, I think you'll be ready.

**Comment by aleaverfay** [comment]
@fdimaio The codecov report isn't showing up for this PR. Is that hook broken? Could you have a look?

File: tmol/score/omega/params.py
Comment (jflat06) [reply tmol/score/omega/params.py:36]: Doesn't the value K need to be converted to a tensor somewhere?

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (jflat06) [reply tmol/score/omega/potentials/omega_pose_score.impl.hh:115]: Is the 32-threads-per-CTA going to be spmething that is energy term specific? Do we want this centrally defined so that we can update it more easily depending on the backend hardware?

File: tmol/score/omega/potentials/compiled.ops.cpp
Comment (jflat06) [reply tmol/score/omega/potentials/compiled.ops.cpp:180]: Good catch! This was intended to be temporary, but somehow slipped by when I self-reviewed. I'll try to add more explicit TODOs in the future.

**Comment by fdimaio** [comment]
> @fdimaio The codecov report isn't showing up for this PR. Is that hook broken? Could you have a look?

Looking now

File: tmol/database/scoring/__init__.py
Comment (jflat06) [reply tmol/database/scoring/__init__.py]: I've rearranged all of the entries so that everything is now in alphabetical order.

File: tmol/score/omega/potentials/omega_pose_score.impl.hh
Comment (jflat06) [reply tmol/score/omega/potentials/omega_pose_score.impl.hh]: I have moved this to tmol/score/common/uaid_util.hh

**Comment by jflat06** [comment]
> This is looking really good. After you've cleaned up the linting (let me know if any part of that process is unclear), the handful of small changes I suggested, and fixed the "remove before merging" cudaDeviceSynchronize call, I think you'll be ready.

Hey Andrew, I am assuming you're talking about the buildkite/tmol/microscope-linting check that is failing? I am unfamiliar with buildkite. I have made an account - do you need to add me to an orangization?

File: tmol/score/omega/omega_energy_term.py
Comment (jflat06) [reply tmol/score/omega/omega_energy_term.py:40]: Something that occurred to me: what do we want to do for types that don't have omega defined? Should we fill the omega_quad_uaids for that block type as all -1? That should cause the uaid resolution function to properly skip this.

Comment (jflat06) [reply tmol/score/omega/omega_energy_term.py:40]: Another quick question - does it make sense to extract the omega values out here, when they are already members of RefinedResidueType? They exist in the torsion_to_uaids Mapping as a Tuple instead of a numpy type, and only exist conditionally. Not sure where to draw the line at redundancy.

File: tmol/score/omega/params.py
Comment (jflat06) [reply tmol/score/omega/params.py:34]: I'm inclined to leave this, so that it can serve as a more general example of how to include parameters for anyone wishing to write a score term in the future. If you'd prefer it be simplified, I can change it.

**Comment by jflat06** [comment]
@aleaverfay - It seems like the only remaining issue is the codecov check, which is failing after the last commit: 

https://app.codecov.io/gh/uw-ipd/tmol/pull/250/commits

This was an attempt to alphabetize the benchmark tests. Is this a concern? Should I revert it?

**Comment by jflat06** [comment]
@aleaverfay @fdimaio Looks like it passed!

**Comment by aleaverfay** [comment]
> @aleaverfay @fdimaio Looks like it passed!

Fantastic! It looks like you'll have to resolve a conflict with trunk, but once you've done that and the tests come back clean, go ahead and squash merge. When you do that, write a solid description of your changes and make sure that the PR number is included in your commit message so that github will link it back to this branch. You can delete the branch once you've merged (I have been not good about that, personally).

**Comment by jflat06** [comment]
Looks like it is blocking the merge since it is still flagged as 'changes requested'. @aleaverfay if things look good, do you want to approve it?

**Comment by jflat06** [comment]
> > Looks like it is blocking the merge since it is still flagged as 'changes requested'. @aleaverfay if things look good, do you want to approve it?
> 
> It looks like linting is still failing (code formatting issues). Possible that you are running the wrong version of black or that the pre-commit hooks are not updating the failing file?

That was a problem previously when I didn't have the conda environment activated by accident. I'll double check that it is correct tomorrow.

**Comment by jflat06** [comment]
Ah, I see what happened. I resolved the merge conflicts directly on github, which obviously didn't run the pre-commit steps. I'll make a note not to do that on this repo.

**Review by fdimaio — APPROVED** [review]
LGTM

File: tmol/score/common/uaid_util.hh
Comment (jflat06) [inline tmol/score/common/uaid_util.hh:29]: @aleaverfay I just noticed while searching for omega that this rogue argument made its way into this function. What's the best way to remove this? Should I make a mini-PR? Or do you just want to update this when you use this function for your task that needed it?

**Pull request merged**
